### PR TITLE
[WIP][virtualization-testing] VM snapshot using virsh and virt-manager

### PR DESCRIPTION
**Pull request by guoxuguang**
Creating and Managing Snapshots
VM Guest snapshots are snapshots of the complete virtual machine including the state of CPU, RAM, devices, and the content of all writable disks. To use virtual machine snapshots, all the attached hard disks need to use the qcow2 disk image format, and at least one of them needs to be writable.

- Verification run: 
gi-guest_developing-on-host-developing-kvm
http://10.67.18.132/tests/100

### CODE REVIEW

**Comment by guoxuguang** [comment]
@mnowaksuse @waynechen55 @Julie-CAO please help me review for this PR

**Comment by Julie-CAO** [comment]
The record_info makes the screenshot help the test results earlier to read.
Is the test only for virsh tool, not for virt-manager?

Throw two questions here and welcome discussion:
One snapshot is created in the testcase, do you think creating more snapshots, at least two snapshots, are more helpful to expose product issues? 
Are we expected to support all the virsh options illustrated in the SUSE guide, such as creating a snapshot with a given name, creating an external snapshot, specifying the memory and etc?
In short, for these QAM tests, our tests are supported to be as simple to run the basic function, or as complete to cover the scenarios in the SUSE document?

**Comment by guoxuguang** [comment]
@mnowaksuse @waynechen55 @Julie-CAO please help me review for this PR

> The record_info makes the screenshot help the test results earlier to read.
> Is the test only for virsh tool, not for virt-manager?
This test support both on virsh and virt-manager together. 
As the first step, just only finish virsh tool (high priority)
> 
> Throw two questions here and welcome discussion:
> One snapshot is created in the testcase, do you think creating more snapshots, at least two snapshots, are more helpful to expose product issues?
Good question, we can add some more basic vm admin here. 
So, before vm admin , we can create first vm snapshot , and after vm admin, we can create sec vm snapshot . 
This PR just only focus on VM snapshot using virsh (create - restore - delete), basic feature. 
Your idea is perfect, we can do this future. 
> Are we expected to support all the virsh options illustrated in the SUSE guide, such as creating a snapshot with a given name, creating an external snapshot, specifying the memory and etc?
This PR just cover Internal snapshots. 
> In short, for these QAM tests, our tests are supported to be as simple to run the basic function, or as complete to cover the scenarios in the SUSE document?
imho, we need to be as simple to run the basic function now, and then we'd better to cover  the scenarios in the SUSE document step by step with QAM together.

File: products/sle/main.pm [hunk 1]
```diff
@@ -754,6 +754,10 @@ elsif (get_var("VIRT_AUTOTEST")) {
     }
     if (get_var("VIRT_PRJ1_GUEST_INSTALL")) {
         loadtest "virt_autotest/guest_installation_run";
+        #Snapshots are supported on KVM VM Host Servers only
+        return unless (check_var('SYSTEM_ROLE', 'kvm'));
```
Comment (ghost) [inline products/sle/main.pm:758]: Outer brackets are redundant.

File: products/sle/main.pm [hunk 2]
```diff
@@ -754,6 +754,10 @@ elsif (get_var("VIRT_AUTOTEST")) {
     }
     if (get_var("VIRT_PRJ1_GUEST_INSTALL")) {
         loadtest "virt_autotest/guest_installation_run";
+        #Snapshots are supported on KVM VM Host Servers only
```
Comment (ghost) [inline products/sle/main.pm:757]: Space after `#`, please.

File: tests/virt_autotest/virsh_snapshot.pm [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+# Summary: Test VM snapshot using virsh (create - restore - delete)
+# Maintainer: Leon Guo <xguo@suse.com>
+ 
+use strict;
+use warnings;
+use testapi;
+use base "virt_autotest_base";
+
+sub run {
+
```
Comment (ghost) [inline tests/virt_autotest/virsh_snapshot.pm:25]: Drop this blank line, please.

File: tests/virt_autotest/virsh_snapshot.pm [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    my $guest = script_output("virsh list --all| sed -n 3p | awk {'print \$2'}");
+
+    record_info "List", "Guest list";
+    assert_script_run "virsh list --all | grep $guest";
+
+    record_info "Create", "snapshot-create";
+    assert_script_run "virsh snapshot-create-as --domain $guest";
```
Comment (ghost) [inline tests/virt_autotest/virsh_snapshot.pm:32]: You either miss the snapshot name attribute, or want to use `snapshot-create` command instead.

File: tests/virt_autotest/virsh_snapshot.pm [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    record_info "snapshot-list", "snapshot-list";
+    assert_script_run "virsh snapshot-list --domain $guest";
```
Comment (ghost) [inline tests/virt_autotest/virsh_snapshot.pm:35]: Here, and at other places, `--domain` does not seems to be an actual option, the synopsis in `virsh(1)` says: `snapshot-list domain` (not `--domain`). 

Edit: Or the least `--domain` is redundant.

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
Maybe you intended to create the PR agains os-autoinst/os-autoinst-distri-opensuse project instead?

It's a good start and the right approach in my mind. I suggest you broaden the test by not just checking that `assert_script_run` returns zero, but `grep` for actual important information.

E.g.: Create snapshot of a given unique name, check that the snapshot is listed for the VM, remove specifically that snapshot, check that the snapshot is done, etc.

**Comment by ghost** [comment]
> Are we expected to support all the virsh options illustrated in the SUSE guide, such as creating a snapshot with a given name, creating an external snapshot, specifying the memory and etc?

Well, to some reasonable degree we should follow SUSE guides. But in some cases we may found out those guides too "hairy" to implement in tests.

> In short, for these QAM tests, our tests are supported to be as simple to run the basic function, or as complete to cover the scenarios in the SUSE document?
> imho, we need to be as simple to run the basic function now, and then we'd better to cover  the scenarios in the SUSE document step by step with QAM together.

I understand this approach as making sure that what is described in guide actually works. And as those guides try to solve real-world issues, I think we should go with complex scenarios, but in small self-contained steps.

File: products/sle/main.pm (hunk 2 above)
Comment (guoxuguang) [reply products/sle/main.pm:757]: Get it, will update PR with this change as your mentioned.

File: products/sle/main.pm (hunk 1 above)
Comment (guoxuguang) [reply products/sle/main.pm:758]: Get it, will update PR with this change as your mentioned too.

File: tests/virt_autotest/virsh_snapshot.pm (hunk 3 above)
Comment (guoxuguang) [reply tests/virt_autotest/virsh_snapshot.pm:25]: Get it, will update PR with this change as your mentioned.

File: tests/virt_autotest/virsh_snapshot.pm (hunk 4 above)
Comment (guoxuguang) [reply tests/virt_autotest/virsh_snapshot.pm:32]: Got it, will try to use snapshot-create command.

File: tests/virt_autotest/virsh_snapshot.pm (hunk 5 above)
Comment (guoxuguang) [reply tests/virt_autotest/virsh_snapshot.pm:35]: refer to https://www.suse.com/documentation/sles-12/singlehtml/book_virt/book_virt.html#sec.libvirt.managing.snapshots, there was used virsh snapshot-list --domain (not domain)

**Comment by cachen** [comment]
2 important features are missing in your coverage refers to the guide. 
- 10.5.3.2 Creating External Snapshots
- 10.5.3.4 Starting a Snapshot: --snapshotname to specify a snapshot name

I strongly suggest you DON'T missing any Mandatory or important parameters/options(such as Michal pointed --name) .
Agree with Michal, The book_virt give a good guidance, but before the code implemented I suggest you manually follow the guide to making sure the guide is correct. And the internal testcase can be restructured.

For example
*virsh for internal snapshot
-create
-management
-start
-delete
*virsh for external snapshot
-create
-management
-start
-delete

File: products/sle/main.pm [hunk 6]
```diff
@@ -754,6 +754,10 @@ elsif (get_var("VIRT_AUTOTEST")) {
...
+        return unless (check_var('SYSTEM_ROLE', 'kvm'));
+        # VM snapshot via virsh
+        loadtest "virt_autotest/virsh_snapshot";
     }
```
Comment (alice-suse) [inline products/sle/main.pm:761]: Please hold on integration of this test to prj1, since what wayne is doing now as clue scripts before calling the scripts  that you and qam newly create, is still wip.
All tasks in the backlog that qam and us share has the same position as clarified in yesterday's dns discussion meeting.

File: tests/virt_autotest/virsh_snapshot.pm [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+ 
+use strict;
+use warnings;
+use testapi;
+use base "virt_autotest_base";
+
+sub run {
+
+    my $guest = script_output("virsh list --all| sed -n 3p | awk {'print \$2'}");
+
```
Comment (alice-suse) [inline tests/virt_autotest/virsh_snapshot.pm:27]: As yesterday mentioned, please do it the way similar to how qam tests. Loop all the guests for the test operations, just like https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/virtualization/xen/hotplugging.pm. Both QAM and us can save efforts when integrating each other if we follow similar style.
